app/models/life_cycle.rb

```diff
@@ -0,0 +1,47 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class LifeCycle < ApplicationRecord
+  has_many :project_life_cycles, class_name: "Project::LifeCycle", dependent: :destroy
+  has_many :projects, through: :project_life_cycles
+
+  belongs_to :color, optional: false
+
+  validates :name, presence: true
+  validates :type, inclusion: { in: %w[Stage Gate], message: :must_be_a_stage_or_gate }
+
+  def initialize(*args)
+    if instance_of? LifeCycle
+      # Do not allow directly instantiating this class
+      raise NotImplementedError, "Cannot instantiate the base LifeCycle class directly. " \
+                                 "Use Stage or Gate instead."
+    end
+
+    super
+  end
```

Comment:
Unfortunately the STI mechanism is not compatible with abstract classes.
Defining an abstract parent class will tell rails that there is no table associated with it, and it disables the STI. Even if you define `self.abstract_class = false` for the child classes, the STI mechanism will not be able to infer the common table name. Defining the `table_name` in each child won't work either, because STI mechanism won't kick in and the `type` field won't be correctly utilised for inferring child classes.

Reply:
Yes, you are correct.
Then it should be enough to validate presence of type, currently validation of inclusion is doing it, but as rails will already check that type is a name of a subclass, I think we don't need to repeat it.